### DISCUSSION TITLE
MTest::computeStiffnessMatrixAndResidual : behaviour intregration failed

### DISCUSSION THREAD

**Discussion by kkolari** [thelfer/tfel#388] [General]
Hi,
I have a quite simple cohesive damage model. 
If I do a small chance into damage function, it has convergence problems although it is not diverging:
- damage (yield) condition = very small
- stresses are in equilibrium 
What does the error message mean, what could I do to solve the problem?
Thanks for adwance!

error message:
The behaviour 'vetosaro' has thrown an CastemException : behaviour integration failed : integration failed for behaviour 'vetosaro'
MTest::computeStiffnessMatrixAndResidual : behaviour intregration failed
Execution failed (GenericSolver::execute: maximum number of sub stepping reached)
-number of period:     36
-number of iterations: 80
-number of sub-steps:  1
Result of test 'unit behaviour test' of group 'MTest'

**Comment by thelfer** [comment]
@kkolari It means that there is a problem when integrating the behavior. Can you share your files ?

**Reply by kkolari** [reply]
I am afraid that I cannot share files yet, but the attached figure explain the problem.

- the picture on the left shows the response with the original damage function
- the picture on the right shows the behavior when a constant value is added.  
- functions are on on the top of the figures
![Mfront error](https://user-images.githubusercontent.com/70758184/231958792-35daacc3-ca30-47e4-8def-3dba9ca0725d.jpg)

**Reply by thelfer** [reply]
You shall have written:

```cxx
feN = SNtrial-EN*deN - st- 5000*(eN+deN);
```

In your statement, 5000*eN is a constant, so the derivative dfeN_ddeN was wrong.

**Reply by kkolari** [reply]
Thanks for quick reply. I do understand your point - Clear mistake! thanks a lot!

**Reply by kkolari** [reply]
The new Equation "feN = SNtrial-EN*deN - st- 5000*(eN+deN);" leads to the same convergence error. 
If tolerance is changed from @Epsilon 1.e-11; to @Epsilon 1.e-9, then it works well.
But with bigger tolerance  also "feN = SNtrial-EN*deN - st- 5000*(eN);"  works!!
I'll clean the files and share them later.
Thanks for help!

**Reply by thelfer** [reply]
You shall normalize your residual for the convergence tolerance value to be meaningful and your variables to avoid conditioning issues (i.e. variables and residual that do not have the same order of magnitudes).

I would suggest:

```cxx
feN = (SNtrial-EN*eN - st- 5000(eN+deN))/EN;"
```

and modifiy the derivatives appropriatly...

**Reply by kkolari** [reply]
Super Great !! Normalization solved the problem!
Thanks - have a nice weekend!
